doc/source/reference/c-api/array.rst

```diff
@@ -1278,6 +1278,14 @@ New data types
     registered (checked only by the address of the pointer), then
     return the previously-assigned type-number.
 
+.. c:function:: int PyArray_TypeNumFromName( \
+        char const *str)
+
+   Given a string return the type-number for the data-type with that string as
+   the type-object name.
+   Returns NPY_NOTYPE without setting an error if no type can be found.
+   Only works for user-defined data-types.
+
```

Comment:
I think this belongs at the end of the section, which would order these functions alphabetically and also make the helper function used after registering the user-defined dtype come after all the registering functions. Also may be worth changing the heading from `New data types` to `User-defined data types`.